File: FlickrAppTests/ExpectationTests.m
//
//  ExpectationTests.m
//  FlickrAppTests
//
//  Created by aSqar on 26.11.2017.
//  Copyright © 2017 Askar Bakirov. All rights reserved.
//

#import <XCTest/XCTest.h>
#import "FeedFetcher.h"
#import "PhotoFetcher.h"
#import "Entities.h"

@interface ExpectationTests : XCTestCase

@end

@implementation ExpectationTests

- (void)setUp {
    [super setUp];
    // Put setup code here. This method is called before the invocation of each test method in the class.
}

- (void)tearDown {
    // Put teardown code here. This method is called after the invocation of each test method in the class.
    [super tearDown];
}

- (void)testExpectation_FeedFetcher
{
    XCTestExpectation *expectation = [self expectationWithDescription:@"FeedFetcher"];
    
    [[FeedFetcher sharedFetcher] fetchManyFromPath:@"" synchronoulsy:YES success:^(NSURLSessionTask *operation, id mappingResult) {
        XCTAssertNotNil(operation, @"mappingResult should not be nil");
        XCTAssertNotNil(mappingResult, @"mappingResult should not be nil");
        
        NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse *)operation.response;
        
        XCTAssertEqual(operation.response.URL.absoluteString, operation.originalRequest.URL.absoluteString, @"HTTP response URL should be equal to original URL");
        XCTAssertEqual(httpResponse.statusCode, 200, @"HTTP response status code should be 200");
        XCTAssertTrue([httpResponse.MIMEType isEqualToString: @"application/json"], @"HTTP response content type should be application/json");
        
        XCTAssertTrue([mappingResult isKindOfClass:[NSArray class]], @"mappingResult should be array");
        NSArray *results = mappingResult;
        XCTAssertTrue(results.count > 0, @"mappingResult should not be empty");
        
        for (Feed *item in results) {
            XCTAssertTrue([item isKindOfClass:[Feed class]], @"element should be of Feed class");
        }
        
        [expectation fulfill];
    } failure:^(NSURLSessionTask *operation, NSError *error) {
        NSLog(@"%@", operation.originalRequest.URL.absoluteString);
        XCTFail(@"Failed with error: %@", error);
    }];
    
    [self waitForExpectations:@[expectation] timeout: 5];
}

- (void)testExpectation_PhotoFetcher
{
    XCTestExpectation *expectation = [self expectationWithDescription:@"PhotoFetcher"];
    
    [[PhotoFetcher sharedFetcher] fetchManyFromPath:@"text=kittens" synchronoulsy:YES success:^(NSURLSessionTask *operation, id mappingResult) {
        XCTAssertNotNil(operation, @"mappingResult should not be nil");
        XCTAssertNotNil(mappingResult, @"mappingResult should not be nil");
        
        NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse *)operation.response;
        
        XCTAssertEqual(operation.response.URL.absoluteString, operation.originalRequest.URL.absoluteString, @"HTTP response URL should be equal to original URL");
        XCTAssertEqual(httpResponse.statusCode, 200, @"HTTP response status code should be 200");
        XCTAssertTrue([httpResponse.MIMEType isEqualToString: @"application/json"], @"HTTP response content type should be application/json");
        
        XCTAssertTrue([mappingResult isKindOfClass:[NSArray class]], @"mappingResult should be array");
        NSArray *results = mappingResult;
        XCTAssertTrue(results.count > 0, @"mappingResult should not be empty");
        
        for (Photo *item in results) {
            XCTAssertTrue([item isKindOfClass:[Photo class]], @"element should be of Feed class");
        }
        
        [expectation fulfill];
    } failure:^(NSURLSessionTask *operation, NSError *error) {
        NSLog(@"%@", operation.originalRequest.URL.absoluteString);
        XCTFail(@"Failed with error: %@", error);
    }];
    
    [self waitForExpectations:@[expectation] timeout: 5];
}

@end

File: FlickrAppTests/ExpectationTests.swift
//
//  ExpectationTests.swift
//  FlickrAppTests
//
//  Created by aSqar on 26.11.2017.
//  Copyright © 2017 Askar Bakirov. All rights reserved.
//

@testable import FlickrApp
import XCTest
import Realm

class ExpectationTests : BaseViewModelTestCase {

    override func setUp() {
        super.setUp()
        // Put setup code here. This method is called before the invocation of each test method in the class.
    }

    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
    }
    
    class MockFeedFetcher : FeedFetcher {
        var _realm:RLMRealm!
        
        init(realm: RLMRealm!)
        {
            _realm = realm
            super.init()
        }
        
        required init() {
            fatalError("init() has not been implemented")
        }
        
        override func realm() -> RLMRealm! {
            return _realm
        }
    }
    
    class MockPhotoFetcher : PhotoFetcher {
        var _realm:RLMRealm!
        
        init(realm: RLMRealm!)
        {
            _realm = realm
            super.init()
        }
        
        required init() {
            fatalError("init() has not been implemented")
        }
        
        override func realm() -> RLMRealm! {
            return _realm
        }
    }

    func testExpectation_FeedFetcher() {
        let expectation:XCTestExpectation! = self.expectation(description: "FeedFetcher")

        let mockFetcher:FeedFetcher! = MockFeedFetcher(realm: self.realm)

        mockFetcher.fetchManyFromPath(restServiceUrl: "", synchronoulsy:true, success:{ (operation,mappingResult) in
            XCTAssertNotNil(operation, "mappingResult should not be nil")
            XCTAssertNotNil(mappingResult, "mappingResult should not be nil")

            let httpResponse:HTTPURLResponse! = operation!.response as! HTTPURLResponse

            XCTAssertEqual(operation!.response!.url!.absoluteString, operation!.originalRequest!.url!.absoluteString, "HTTP response URL should be equal to original URL")
            XCTAssertEqual(httpResponse.statusCode, 200, "HTTP response status code should be 200")
            XCTAssertTrue((httpResponse.mimeType == "application/json"), "HTTP response content type should be application/json")

            XCTAssertTrue((mappingResult is NSArray), "mappingResult should be array")
            let results:[AnyObject]! = mappingResult as! [AnyObject]
            XCTAssertTrue(results.count > 0, "mappingResult should not be empty")

            for item:Feed! in results as! [Feed]!{
                XCTAssertTrue((item != nil), "element should be of Feed class")
             }

            expectation.fulfill()
        }, failure:{ (operation,error) in
            NSLog("%@", operation!.originalRequest!.url!.absoluteString)
            //XCTFail();
        })

        self.wait(for: [expectation], timeout: 5)
    }

    func testExpectation_PhotoFetcher() {
        let expectation:XCTestExpectation! = self.expectation(description: "PhotoFetcher")

        let mockFetcher:PhotoFetcher! = MockPhotoFetcher(realm: self.realm)

        mockFetcher.fetchManyFromPath(restServiceUrl: "text=kittens", synchronoulsy:true, success:{ (operation,mappingResult) in
            XCTAssertNotNil(operation, "mappingResult should not be nil")
            XCTAssertNotNil(mappingResult, "mappingResult should not be nil")

            let httpResponse:HTTPURLResponse! = operation!.response as! HTTPURLResponse

            XCTAssertEqual(operation!.response!.url!.absoluteString, operation!.originalRequest!.url!.absoluteString, "HTTP response URL should be equal to original URL")
            XCTAssertEqual(httpResponse.statusCode, 200, "HTTP response status code should be 200")
            XCTAssertTrue((httpResponse.mimeType == "application/json"), "HTTP response content type should be application/json")

            XCTAssertTrue((mappingResult is NSArray), "mappingResult should be array")
            let results:[AnyObject]! = mappingResult as! [AnyObject]
            XCTAssertTrue(results.count > 0, "mappingResult should not be empty")

            for item:Photo! in results as! [Photo]! {
                XCTAssertTrue((item != nil), "element should be of Feed class")
             }

            expectation.fulfill()
        }, failure:{ (operation,error) in
            NSLog("%@", operation!.originalRequest!.url!.absoluteString)
            XCTFail()
        })

        self.wait(for: [expectation], timeout: 5)
    }
}
